### DISCUSSION TITLE
Persistent contexts and projects

### DISCUSSION THREAD

**Discussion by eamonburns** [todotxt/todo.txt-cli#428] [General]
Is there a way to have contexts and projects that are always listed in `listproj` and `listcon`, even if there are no todos that currently have them? I have a script that allows me to easily pick between available contexts and projects, and it would be nice to not have to retype them every time I have completed all the tasks currently related to them

**Comment by inkarkat** [comment]
To add additional projects or contents to `listproj` / `listcon`, you can override the built-in commands (or add alternative commands) that just delegate to the built-in command and add more output on their own:

```
#!/bin/bash

[ "$1" = "usage" ] && TODO_ACTIONS_DIR=/ exec "$TODO_FULL_SH" help listcon  # Delegate help to the built-in command, avoid endless recursion by clearing the actions dir.
shift

echo $'@static-context1\n@static-context2'  # The static list of additional contexts.
listWordsWithSigil '@' "$@" # Delegate to the built-in function.
```

However, that means maintaining the static list within the script (or a separate file). I think a more useful implementation would be looking into the done tasks; the Bash completion [does that already](https://github.com/inkarkat/todo.txt-cli/blob/2d70a0aadfca885ec49c734ab56d19cd1224ff31/todo_completion?plain=1#L50-L51) when no project / context is found.

`TODOTXT_SOURCEVAR=\$DONE_FILE todo.sh listcon ...` looks for contexts in done files; if you combine that with the regular `listcon` output, de-duplicate via `sort --unique`, you'll get a list of all current and past contexts.

**Reply by eamonburns** [reply]
Ok, I will probably make my own custom action then
